File: src/main/java/com/durhack/sharpshot/nodes/ConstantNode.java
package com.durhack.sharpshot.nodes;

import com.durhack.sharpshot.Bullet;
import com.durhack.sharpshot.Direction;
import com.durhack.sharpshot.INode;
import com.durhack.sharpshot.gui.Triangle;
import javafx.scene.Node;
import javafx.scene.paint.Color;
import org.jetbrains.annotations.NotNull;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

public class ConstantNode implements INode {
    private Direction dir = Direction.UP;
    private BigInteger value;

    public ConstantNode(BigInteger value) {
        this.value = value;
    }

    public void setValue(BigInteger value) {
        this.value = value;
    }

    public BigInteger getValue() {
        return value;
    }

    @Override
    public @NotNull Direction getRotation() {
        return dir;
    }

    @Override
    public void rotateClockwise() {
        dir = Direction.clockwiseOf(dir);
    }

    @Override
    public @NotNull Map<Direction, BigInteger> run(@NotNull Bullet bullet) {
        // Shoot out a bullet
        HashMap<Direction, BigInteger> map = new HashMap<>();
        map.put(bullet.getDirection(), bullet.getValue());
        map.put(Direction.UP, getValue()); // if both same direction, only constant comes out
        return map;
    }

    @Override
    public @NotNull Node toGraphic() {
        return new Triangle(getRotation(), Color.LIMEGREEN, value.toString());
    }

    public void reset() {}

    @Override
    public String toString() {
        return "Constant";
    }
}

File: src/main/java/com/durhack/sharpshot/nodes/ConstantNode.kt
package com.durhack.sharpshot.nodes

import com.durhack.sharpshot.Bullet
import com.durhack.sharpshot.Direction
import com.durhack.sharpshot.gui.shapes.Triangle
import com.durhack.sharpshot.gui.util.getNumberInput
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import javafx.scene.paint.Color
import java.math.BigInteger

class ConstantNode(var value: BigInteger?) : INode() {
    /**
     * If both the bullet and the output are in the same direction, the second element in the map overrides the first
     * so the input bullet is destroyed
     */
    override fun run(bullet: Bullet): Map<Direction, BigInteger?> =
            mutableMapOf(bullet.direction to bullet.value,
                         Direction.UP to value
                        )

    override fun graphic() = Triangle(rotation,
                                                                       Color.LIMEGREEN,
                                                                       value?.toString() ?: "")
    override fun reset() {}

    override val type = "constant"

    override val tooltip = "Whenever a bullet passes through, release another bullet with pre-set value"
    override val factory = {
        val value = getNumberInput("Enter Input Index",
                                                                  "Blank to shoot empty bullet at start\nArguments are 0-indexed")
        if (value.isPresent) {
            val string = value.get()
            when {
                string.isBlank() -> ConstantNode(null)
                else -> ConstantNode(BigInteger(string))
            }
        }
        else {
            null
        }
    }

    override fun toJson(): JsonElement {
        val json = super.toJson().asJsonObject
        json.addProperty("value", value?.toString() ?: "null")
        return json
    }

    override val jsonFactory: (JsonObject) -> ConstantNode = { json ->
        val valueString = json["value"].asString
        val value = when (valueString) {
            "null" -> null
            else -> BigInteger(valueString)
        }

        val node = ConstantNode(value)
        val rotation = json["rotation"].asInt
        node.rotation = Direction.ofQuarters(rotation)
        node
    }
}
